src/image_service.rs

```diff
@@ -1,24 +1,195 @@
+                "registry.docker.io",
+                "https://registry-1.docker.io",
+                "https://auth.docker.io/token",
+            )],
+
+            layer_lock: Default::default(),
+        })
+    }
+
+    /// List already pulled images
+    /// TODO: use passed spec
+    pub fn list_images(&self, _spec: Option<&ImageSpec>) -> Result<Vec<criapi::Image>, ImageError> {
+        use std::io::Cursor;
+
+        let mut k = sled::IVec::default();
+        let mut out = vec![];
+        while let Some((ik, v)) = self.database.get_gt(&k)? {
+            k = ik.clone();
+            out.push(criapi::Image::decode(&mut Cursor::new(&v))?);
+        }
+        Ok(out)
+    }
+
+    pub async fn download_layer(
+        &self,
+        registry: &DockerRegistryClientV2,
+        image: &str,
+        digest: &Digest,
+    ) -> Result<()> {
+        let _lock = self.layer_lock.lock(digest.to_string()).await;
+
+        let mut layer_dir = self.layers.clone();
+        layer_dir.push(&digest.to_string());
+
+        let meta = fs::metadata(&layer_dir).await;
+        if meta.is_ok() && meta?.is_dir() {
+            return Ok(());
+        }
+
+        let tmp_layer_dir = tempfile::tempdir_in(&self.layers)?;
+
+        let out_digest = Arc::new(Mutex::new(None));
+
+        let blob = registry.blob(image, digest).await?;
+        let blob_stream = blob_to_stream(blob, out_digest.clone());
+        let blob_reader = stream_reader(blob_stream);
+        let deflate = Box::pin(GzipDecoder::new(blob_reader));
+        let mut tar = Archive::new(deflate);
+
+        tar.unpack(&tmp_layer_dir).await?;
+
+        // TODO: return error instead of panic on digest mismatch
+        // assert_eq!(out_digest.as_ref().unwrap(), digest);
+
+        fs::rename(tmp_layer_dir.into_path(), layer_dir).await?;
+
+        Ok(())
+    }
+
+    pub async fn pull_image(&self, spec: &ImageSpec) -> Result<Option<Manifest>> {
+        // TODO: handle registry name in image name (I.e docker.io/redis:latest)
+        let parts = spec.image.split(':').collect::<Vec<&str>>();
+        assert_eq!(parts.len(), 2, "bad name format");
+        let image_name = parts[0];
+        let reference = parts[1];
+        for registry in &self.registries {
+            let response = registry.auth("repository", image_name, "pull").await;
+            let mut registry = registry.clone();
+            if let Ok(token) = response {
+                registry.set_auth_token(Some(token));
+            }
+
+            let manifests = match registry.list_manifests(image_name, reference).await {
+                Ok(m) => m,
+                // TODO: abort on error
+                Err(e) => {
+                    println!("Error: {:?}", e);
+                    continue;
+                }
+            };
+            let mut found_manifest = None;
+            for manifest in manifests.manifests {
+                // TODO: check platform with passed annotations
```

Comment:
I guess you meant in case of [index manifest](https://github.com/opencontainers/image-spec/blob/master/image-index.md), we should check against platform and os fields to see if they match.